### PR TITLE
changefeed,kvClient, sink (ticdc): support bidirectional replication

### LINKED ISSUE #7736: ticdc: support bidirectional replication.
### Is your feature request related to a problem?

please check: https://github.com/pingcap/tidb/issues/38587

### Describe the feature you'd like

as above.

### Describe alternatives you've considered

_No response_

### Teachability, Documentation, Adoption, Migration Strategy

_No response_

### PR DESCRIPTION
**Pull request by asddongmen**
### What problem does this PR solve?

Issue Number: close #7736 

### What is changed and how it works?
1. Add `BDRMode` field to `ReplicaConfig` to config changefeed for BDR mode.
2. Add `filterLoop` field to `CDCClient` struct, which will be used in `cdcpb.ChangeDataRequest` to tell the TiKV cdc component to filter `rowChange` that are written by another TiCDC.
3. When changefeed is initializing, it will get `source_id` of the upstream TiDB through PD client, and set it to `Sink` config.

**Note**: 
1. When downstream is TiDB, TiCDC will exctue `SET SESSION tidb_cdc_write_source = source_id;` before commit a txn, which make every txn written by TiCDC has a hidden value `tidb_cdc_write_source = source_id`. This can help TiKV cdc coponent to decide whether it should ignore a rowChange event.
2. If the **dowstream** TiDB does not support `tidb_cdc_write_source` system variable and the user set `bdr-mode = true` for a changefeed, create changefeed request will return an error.
3. ### Check List 

#### Tests 

 - Integration test
 - Manual test (add detailed scripts or steps below)
  See test plan for detail.

### Release note 

```release-note
Support bidirectional replication.
```

### CODE REVIEW

File: cdc/sinkv2/eventsink/txn/mysql/mysql.go [hunk 1]
```diff
@@ -88,6 +89,15 @@ func NewMySQLBackends(
 	if err != nil {
 		return nil, err
 	}
+
+	//Note(dongmen): Only for BDR mode use for now.
+	sysVariables := make(map[string]string)
+	// TODO(dongmen): set a correct value for this variable after the `source_id` design is done.
+	sysVariables["tidb_cdc_write_source"] = strconv.Itoa(1)
```
Comment (nongfushanquan) [inline cdc/sinkv2/eventsink/txn/mysql/mysql.go:96]: where is the source id ?

Comment (asddongmen) [reply cdc/sinkv2/eventsink/txn/mysql/mysql.go]: Added. PTAL

File: cdc/owner/changefeed.go [hunk 2]
```diff
@@ -900,6 +912,15 @@ func (c *changefeed) asyncExecDDLEvent(ctx cdcContext.Context,
 			zap.String("changefeed", c.id.ID), zap.Any("event", ddlEvent))
 		return true, nil
 	}
+
+	// check whether in bdr mode, if so, we need to skip all DDLs
+	if c.state.Info.Config.BDRMode {
+		log.Info("ignore the DDL event in BDR mode",
```
Comment (nongfushanquan) [inline cdc/owner/changefeed.go:918]: DEBUG level?

File: cdc/sinkv2/eventsink/txn/mysql/mysql.go [hunk 3]
```diff
@@ -93,6 +93,12 @@ func NewMySQLBackends(
 	if err != nil {
 		return nil, err
 	}
+
+	cfg.IsTiDB, err = isTiDB(ctx, db)
```
Comment (nongfushanquan) [inline cdc/sinkv2/eventsink/txn/mysql/mysql.go:97]: If the BDR is enabled , can we pre-check whether the downstream is tidb

File: cdc/owner/changefeed.go [hunk 4]
```diff
@@ -501,6 +502,17 @@ LOOP:
 	cancelCtx, cancel := cdcContext.WithCancel(ctx)
 	c.cancel = cancel
 
+	sourceID, err := pdutil.GetSourceID(ctx, c.upstream.PDClient)
```
Comment (nongfushanquan) [inline cdc/owner/changefeed.go:505]: it is not necessary to get sourced if  the BDR is disabled

Comment (asddongmen) [reply cdc/owner/changefeed.go:505]: yes, we always need it.

File: cdc/sinkv2/eventsink/txn/mysql/mysql.go (hunk 3 above)
Comment (asddongmen) [reply cdc/sinkv2/eventsink/txn/mysql/mysql.go:97]: done

File: cdc/owner/changefeed.go (hunk 2 above)
Comment (asddongmen) [reply cdc/owner/changefeed.go:918]: All ddl related log are info level in cdc.

File: cdc/api/v2/model.go [hunk 5]
```diff
@@ -95,6 +95,7 @@ type ReplicaConfig struct {
 	IgnoreIneligibleTable bool              `json:"ignore_ineligible_table"`
 	CheckGCSafePoint      bool              `json:"check_gc_safe_point"`
 	EnableSyncPoint       bool              `json:"enable_sync_point"`
+	BDRMode               bool              `json:"bdr-mode"`
```
Comment (zhaoxinyu) [inline cdc/api/v2/model.go:98]:
```suggestion
	BDRMode               bool              `json:"bdr_mode"`
```

File: pkg/sink/mysql/config.go [hunk 6]
```diff
@@ -80,7 +80,11 @@ type Config struct {
 	TLS                 string
 	ForceReplicate      bool
 	EnableOldValue      bool
-	BatchDMLEnable      bool
+
+	IsTiDB         bool // IsTiDB is true if the downstream is TiDB
+	SourceID       uint64
+	BDRMode        bool
```
Comment (zhaoxinyu) [inline pkg/sink/mysql/config.go:86]: It's better to add some comments about `BDRMode`

File: cdc/sink/validator.go [hunk 7]
```diff
@@ -96,3 +106,63 @@ func preCheckSinkURI(sinkURIStr string) error {
 
 	return nil
 }
+
+func checkBDRMode(ctx context.Context, sinkURI string, replicaConfig *config.ReplicaConfig) error {
+	maskSinkUri, err := util.MaskSinkURI(sinkURI)
+	if err != nil {
+		return err
+	}
+	isDownstreamTiDBOrMySQL, err := checkIsDownstreamTiDBoRMySQL(sinkURI)
```
Comment (zhaoxinyu) [inline cdc/sink/validator.go:115]:
```suggestion
	isDownstreamTiDBOrMySQL, err := checkIsDownstreamTiDBOrMySQL(sinkURI)
```

File: cdc/sink/validator.go [hunk 8]
```diff
@@ -96,3 +106,63 @@ func preCheckSinkURI(sinkURIStr string) error {
...
+	if !supported {
+		// TODO: refine this error message
+		return cerror.ErrSinkURIInvalid.
+			GenWithStack("downstream database does not support BDR mode, "+
+				"please check your config, sink uri: %s", maskSinkUri)
+	}
+	return nil
+}
+
+func checkIsDownstreamTiDBoRMySQL(sinkURI string) (bool, error) {
```
Comment (zhaoxinyu) [inline cdc/sink/validator.go:155]: We already have the same functionality in `IsMySQLCompatibleScheme` in `pkg/sink/sink.go`.

File: pkg/sink/mysql/db_helper.go [hunk 9]
```diff
@@ -266,3 +239,85 @@ func CheckAndAdjustPassword(ctx context.Context, dbConfig *dmysql.Config, dbConn
...
+		dsn.Params["time_zone"] = cfg.Timezone
+	}
+	dsn.Params["readTimeout"] = cfg.ReadTimeout
+	dsn.Params["writeTimeout"] = cfg.WriteTimeout
+	dsn.Params["timeout"] = cfg.DialTimeout
+	return dsn, nil
+}
+
+// CheckIfSupportBDRMode checks if the downstream supports BDR mode.
+func CheckIfSupportBDRMode(ctx context.Context, db *sql.DB) (bool, error) {
```
Comment (zhaoxinyu) [inline pkg/sink/mysql/db_helper.go:283]:
```suggestion
func CheckIfBDRModeIsSupported(ctx context.Context, db *sql.DB) (bool, error) {
```

File: cdc/sinkv2/eventsink/txn/mysql/mysql.go [hunk 10]
```diff
@@ -612,3 +623,23 @@ func getSQLErrCode(err error) (errors.ErrCode, bool) {
...
+		return nil
+	}
+	// downstream is TiDB, set system variables.
+	// We should always try to set this variable, and ignore the error if
+	// downstream does not support this variable, it is by design.
+	query := fmt.Sprintf("SET SESSION %s = %d", "tidb_cdc_write_source", s.cfg.SourceID)
+	_, err := txn.ExecContext(ctx, query)
+	if err != nil {
+		if mysqlErr, ok := errors.Cause(err).(*dmysql.MySQLError); ok &&
+			mysqlErr.Number == mysql.ErrUnknownSystemVariable {
```
Comment (zhaoxinyu) [inline cdc/sinkv2/eventsink/txn/mysql/mysql.go:640]: Should this condition return nil?

File: pkg/sink/mysql/db_helper.go [hunk 11]
```diff
@@ -266,3 +239,85 @@ func CheckAndAdjustPassword(ctx context.Context, dbConfig *dmysql.Config, dbConn
...
+	isTiDB, err := CheckIsTiDB(ctx, db)
+	if err != nil || !isTiDB {
+		return false, err
+	}
+	testSourceID := 1
+	// downstream is TiDB, set system variables.
+	// We should always try to set this variable, and ignore the error if
+	// downstream does not support this variable, it is by design.
+	query := fmt.Sprintf("SET SESSION %s = %d", "tidb_cdc_write_source", testSourceID)
+	//TODO: remove this log after fully testing
```
Comment (zhaoxinyu) [inline pkg/sink/mysql/db_helper.go:293]:
```suggestion
```

File: cdc/sink/validator.go (hunk 8 above)
Comment (asddongmen) [reply cdc/sink/validator.go:155]: thanks

File: cdc/sinkv2/eventsink/txn/mysql/mysql.go (hunk 10 above)
Comment (asddongmen) [reply cdc/sinkv2/eventsink/txn/mysql/mysql.go:640]: yes

File: cdc/kv/client.go [hunk 12]
```diff
@@ -234,6 +234,7 @@ type CDCClient struct {
 	}
 	ingressCommitTs   model.Ts
 	ingressResolvedTs model.Ts
+	filterLoop        bool
```
Comment (3AceShowHand) [inline cdc/kv/client.go:237]: how about add a comment about this field to indicate it's usage ?

File: cdc/owner/changefeed.go [hunk 13]
```diff
@@ -160,11 +161,11 @@ func newChangefeed(
 func newChangefeed4Test(
 	id model.ChangeFeedID, state *orchestrator.ChangefeedReactorState, up *upstream.Upstream,
 	newDDLPuller func(ctx context.Context,
-		replicaConfig *config.ReplicaConfig,
-		up *upstream.Upstream,
-		startTs uint64,
-		changefeed model.ChangeFeedID,
-	) (puller.DDLPuller, error),
+	replicaConfig *config.ReplicaConfig,
```
Comment (3AceShowHand) [inline cdc/owner/changefeed.go:164]: This looks changed by formating, can we avoid it?

File: go.mod [hunk 14]
```diff
@@ -295,3 +295,5 @@ replace github.com/benbjohnson/clock v1.3.0 => github.com/benbjohnson/clock v1.1
 
 // copy from TiDB
 replace go.opencensus.io => go.opencensus.io v0.23.1-0.20220331163232-052120675fac
+
+replace github.com/pingcap/kvproto => github.com/xiongjiwei/kvproto v0.0.0-20221116081748-24d900443289
```
Comment (3AceShowHand) [inline go.mod:299]: when the `kvproto` is up to date, this `replace` should be removed.

**Review by 3AceShowHand — APPROVED** [review]

File: go.mod (hunk 14 above)
Comment (asddongmen) [reply go.mod:299]: yes, thanks for your reminder

File: cdc/kv/client.go (hunk 12 above)
Comment (asddongmen) [reply cdc/kv/client.go:237]: sure

**Review by nongfushanquan — APPROVED** [review]

File: cdc/sinkv2/eventsink/txn/mysql/mysql.go [hunk 15]
```diff
@@ -541,12 +547,17 @@ func (s *mysqlBackend) execDMLWithMaxRetries(pctx context.Context, dmls *prepare
 				}
 			}
 
+			// we set write source for each txn,
+			// so we can use it to trace the data source
+			if err = s.setWriteSource(ctx, tx); err != nil {
+				return 0, err
```
[the remark this answers is not in the document]
Comment (asddongmen) [reply cdc/sinkv2/eventsink/txn/mysql/mysql.go:564]: done

**Pull request merged**